src/Riok.Mapperly/MapperResults.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+
+namespace Riok.Mapperly;
+
+public readonly struct MapperResults : IEquatable<MapperResults>
+{
+    public static readonly MapperResults Empty = new(ImmutableArray<MapperNode>.Empty, ImmutableArray<Diagnostic>.Empty);
+
+    public MapperResults(ImmutableArray<MapperNode> mappers, ImmutableArray<Diagnostic> diagnostics)
+    {
+        Mappers = mappers;
+        Diagnostics = diagnostics;
+    }
+
+    public ImmutableArray<MapperNode> Mappers { get; }
+    public ImmutableArray<Diagnostic> Diagnostics { get; }
+
+    public bool Equals(MapperResults other)
+    {
+        return Mappers.SequenceEqual(other.Mappers) && Diagnostics.SequenceEqual(other.Diagnostics);
+    }
+
+    public override bool Equals(object? obj) => obj is MapperResults other && Equals(other);
+
+    public override int GetHashCode()
+    {
+        unchecked
+        {
+            var hash = 0;
+            foreach (var mapper in Mappers)
+            {
+                hash = Combine(hash, mapper.GetHashCode());
+            }
+            foreach (var diagnostic in Diagnostics)
+            {
+                hash = Combine(hash, diagnostic.GetHashCode());
+            }
+            return hash;
+        }
+
+        static int Combine(int h1, int h2)
```

Comment:
We could probably backport the [HashCode](https://learn.microsoft.com/en-us/dotnet/api/system.hashcode) struct?